### PR TITLE
fix(compiler): don't convert undefined to null literals

### LINKED ISSUE #11493: `null === undefined` and similar expressions evaluate to true in the template compiler.
**I'm submitting a ...**  (check one with "x")

```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior** 
`variable === undefined` is true when `variable` is `null`, and `null === undefined` is true in templates.
`variable === null` is false (as expected) when `variable` is `undefined`, but `undefined === null` is true.

**Expected behavior**
`variable === undefined` should evaluate to false when `variable` is `null`, and `null === undefined` should evaluate to false.
`variable === null` should evaluate to false (as it does) when `variable` is `undefined`, and `undefined === null` should evaluate to false.
**Reproduction of the problem**
http://plnkr.co/edit/4QQdILnqyswJAXMIPyHK?p=preview

**What is the motivation / use case for changing the behavior?**
It violates common sense in JavaScript and TypeScript to an offensive degree.
- **Angular version:** 2.0.0-rc.6
- **Browser:** all 
- **Language:** all

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
Fixes #11493

### CODE REVIEW

**Comment by vicb** [comment]
LGTM haven't done the change yesterday because @tbosch thinks we need more tests (g3) before doing the change.

Let's land this after final.

Thanks

File: modules/@angular/compiler/src/output/abstract_emitter.ts [hunk 1]
```diff
@@ -250,13 +250,10 @@ export abstract class AbstractEmitterVisitor implements o.StatementVisitor, o.Ex
     ctx.print(`)`);
     return null;
   }
-  visitLiteralExpr(ast: o.LiteralExpr, ctx: EmitterVisitorContext, absentValue: string = 'null'):
```
Comment (vicb) [inline modules/@angular/compiler/src/output/abstract_emitter.ts]: `absentValue` is a hack for type checking

Comment (vicb) [inline modules/@angular/compiler/src/output/abstract_emitter.ts]: seems like you haven't updated the call site

File: modules/@angular/compiler/src/output/ts_emitter.ts [hunk 2]
```diff
@@ -72,7 +72,12 @@ class _TsEmitterVisitor extends AbstractEmitterVisitor implements o.TypeVisitor
   }
 
   visitLiteralExpr(ast: o.LiteralExpr, ctx: EmitterVisitorContext): any {
-    super.visitLiteralExpr(ast, ctx, '(null as any)');
+    var value = ast.value;
```
Comment (vicb) [inline modules/@angular/compiler/src/output/ts_emitter.ts]: const

File: modules/@angular/core/test/linker/integration_spec.ts [hunk 3]
```diff
@@ -61,6 +61,17 @@ function declareTests({useJit}: {useJit: boolean}) {
         expect(fixture.nativeElement).toHaveText('');
       });
 
+      it('should allow both null and undefined in expressions', () => {
+        TestBed.configureTestingModule({declarations: [MyComp]});
```
Comment (vicb) [inline modules/@angular/core/test/linker/integration_spec.ts]:
```ts
const template='...';
const fixture = tB.x.z.y.create();
```

File: modules/@angular/compiler/src/output/abstract_emitter.ts (hunk 1 above)
Comment (pkozlowski-opensource) [inline modules/@angular/compiler/src/output/abstract_emitter.ts]: Yep, but with the change in this PR it is not needed anymore as the type-specific code goes to TS-specific class.

File: modules/@angular/core/test/linker/integration_spec.ts (hunk 3 above)
Comment (pkozlowski-opensource) [inline modules/@angular/core/test/linker/integration_spec.ts]: Huh? Not sure I got you here....

File: modules/@angular/compiler/src/output/abstract_emitter.ts (hunk 1 above)
Comment (vicb) [inline modules/@angular/compiler/src/output/abstract_emitter.ts]: question is do we really need this hack ?

File: modules/@angular/core/test/linker/integration_spec.ts (hunk 3 above)
Comment (vicb) [inline modules/@angular/core/test/linker/integration_spec.ts]: refactor, remove TB duplication by using method chaining

Comment (pkozlowski-opensource) [inline modules/@angular/core/test/linker/integration_spec.ts]: Oh, neat! We should probably do the same for other tests in this file (don't want to send a massive diff in this PR but can prepare a separate one if needed)

Comment (vicb) [inline modules/@angular/core/test/linker/integration_spec.ts]: I've done this for common tests.

We probably need to do a cleanup pass on the tests after final & extract common patterns into helpers and may be add them to the testing lib

**Comment by tbosch** [comment]
Side note: G3 is using strict null checks. So we need to make sure the generated code complies and does not break G3.
Otherwise looks good to me too.

**Comment by vicb** [comment]
we should also address https://github.com/angular/angular/pull/11846 when g3 has migrated to TS 2 - should probably be part of the same effort

**Comment by vicb** [comment]
@pkozlowski-opensource could you please rebase this PR.

We should be able to merge this now that TS in g3 has been updated.

/cc @tbosch @alexeagle

**Comment by pkozlowski-opensource** [comment]
@vicb rebased

**Pull request merged**